net/probe-cli/patches-probe-engine/001-fix-cgo.patch

```diff
@@ -0,0 +1,21 @@
+--- a/measurementkit/task_cgo.go
++++ b/measurementkit/task_cgo.go
+@@ -38,6 +38,18 @@ import (
+ 	// #cgo linux,amd64 LDFLAGS: /usr/lib/libevent_extra.a
+ 	// #cgo linux,amd64 LDFLAGS: /usr/lib/libevent_pthreads.a
+ 	// #cgo linux,amd64 LDFLAGS: /lib/libz.a
++	//
++	// #cgo LDFLAGS: -lmeasurement_kit
++	// #cgo LDFLAGS: -lmaxminddb
++	// #cgo LDFLAGS: -lcurl
++	// #cgo LDFLAGS: -lnghttp2
++	// #cgo LDFLAGS: -levent_openssl
++	// #cgo LDFLAGS: -lssl
++	// #cgo LDFLAGS: -lcrypto
++	// #cgo LDFLAGS: -levent_core
++	// #cgo LDFLAGS: -levent_extra
++	// #cgo LDFLAGS: -levent_pthreads
++	// #cgo LDFLAGS: -lz
```

Comment:
Can this patch be upstreamed?

Reply:
I created an [issue](https://github.com/ooni/probe-engine/issues/267) , but it's not their priority right now.

Reply:
Would it work if you added these flags to `TARGET_LDFLAGS` instead of patching this file?

Reply:
I tried that, but it failed in the liking phase. arch Linux has[ similar patch](https://aur.archlinux.org/cgit/aur.git/tree/ooni-engine-measurementkit-task_gco.go.patch?h=ooni-probe-cli-git) 

https://github.com/ooni/probe-engine/blob/master/measurementkit/task_cgo.go#L29

Reply:
I've added `CGO_LDFLAGS` in #11255. That should allow you to set `TARGET_LDFLAGS` and have it affect cgo compilation (instead of having to patch the flags into the file).

I suggest patching out the `#cgo linux,amd64 LDFLAGS:` lines though, since that will affect x86-64 builds.

Reply:
If you add the flags to `TARGET_LDFLAGS` instead of patching them into the file, does it still not work?

Reply:
Yes, it works but for some reason, probe-cli now returns for test **runtime error: invalid memory address or nil pointer dereference**. I'm not sure why. So I changed PR to WIP

Reply:
How do I test probe-cli to get this error?

Reply:
```
root@turris:~# ooniprobe run 
   • Running all tests
   ⨯ panic in app.Run          error=runtime error: invalid memory address or nil pointer dereference
   • sending exception backtrace

```

Reply:
Could this be the panic?

```
runtime.errorString: runtime error: invalid memory address or nil pointer dereference
  File "runtime/panic.go", line 199, in panicmem
  File "runtime/signal_unix.go", line 394, in sigpanic
  File "runtime/internal/atomic/asm_arm.s", line 185, in armLoad64
  File "github.com/ooni/probe-cli@/ooni.go", line 45, in IsTerminated
  File "github.com/ooni/probe-cli@/internal/cli/run/run.go", line 51, in runNettestGroup
  File "github.com/ooni/probe-cli@/internal/cli/run/run.go", line 160, in func7
  File "github.com/alecthomas/kingpin@v2.2.6+incompatible/actions.go", line 28, in applyActions
  File "github.com/alecthomas/kingpin@v2.2.6+incompatible/app.go", line 557, in applyActions
  File "github.com/alecthomas/kingpin@v2.2.6+incompatible/app.go", line 390, in execute
  File "github.com/alecthomas/kingpin@v2.2.6+incompatible/app.go", line 222, in Parse
  File "github.com/ooni/probe-cli@/internal/cli/app/app.go", line 14, in Run
  File "github.com/getsentry/raven-go@v0.0.0-20190419175539-919484f041ea/client.go", line 821, in CapturePanic
  File "github.com/ooni/probe-cli@/internal/crashreport/crashreport.go", line 20, in CapturePanic
  File "github.com/ooni/probe-cli@/cmd/ooniprobe/main.go", line 23, in main
  File "runtime/proc.go", line 203, in main
```

Reply:
I have no experience with sentry. but here is verbose version of that call 
```
root@turris:/# ooniprobe -v run im
   • ooni version 3.0.0-rc.9  
   • Reading default config file
   • Connecting to database sqlite3:///root/.ooniprobe/db/main.sqlite3
   • running migrations       
   • performed 0 migrations   
   • Using collector:         
   • Using bouncer:           
[engine] iplookup: using avast
[engine] jsonapi: method: GET
[engine] jsonapi: URL: https://ip-info.ff.avast.com/v1/info
[engine] dns: resolving ip-info.ff.avast.com
[engine] dns: got 2 entries
[engine] - 5.62.40.217
[engine] - 5.62.40.218
[engine] connect: using tcp, 5.62.40.217:443
[engine] connect: connected to tcp, 5.62.40.217:443
[engine] tls: starting handshake
[engine] tls: handshake OK
[engine] - negotiated protocol: 
[engine] - version: TLSv1.2
[engine] http: connection to 5.62.40.217:443 ready; sending request
[engine] > GET /v1/info HTTP/1.1
[engine] > User-Agent: ooniprobe-cli/3.0.0-rc.9
[engine] > Host: ip-info.ff.avast.com
[engine] > Accept-Encoding: gzip
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Content-Type: application/json
[engine] < Content-Length: 355
[engine] < Connection: keep-alive
[engine] < Server: nginx
[engine] < Date: Wed, 12 Feb 2020 10:55:50 GMT
[engine] <
[engine] jsonapi: response body: {"ip":"94.112.143.156","continent":"Europe","continentCode":"EU","country":"CZ","countryName":"Czechia","subdivisions":["64","642"],"city":"Brno","timezone":"Europe/Prague","latitude":49.2012,"longitude":16.635,"isp":"UPC Ceska Republika","asnNumber":6830,"asnOrganization":"Liberty Global B.V.","organization":"UPC Ceska Republika","postalCode":"612 00"}
[engine] iplookup: IP: 94.112.143.156
[engine] mmdblookup: ASN: &{AutonomousSystemNumber:6830 AutonomousSystemOrganization:Liberty Global B.V.}
[engine] mmdblookup: Country: &{Continent:{Code:EU GeoNameID:6255148 Names:map[de:Europa en:Europe es:Europa fr:Europe ja:ヨーロッパ pt-BR:Europa ru:Европа zh-CN:欧洲]} Country:{GeoNameID:3077311 IsInEuropeanUnion:true IsoCode:CZ Names:map[de:Tschechien en:Czechia es:República Checa fr:République tchèque ja:チェコ共和国 pt-BR:Tchéquia ru:Чешская Республика zh-CN:捷克共和国]} RegisteredCountry:{GeoNameID:3077311 IsInEuropeanUnion:true IsoCode:CZ Names:map[de:Tschechien en:Czechia es:República Checa fr:République tchèque ja:チェコ共和国 pt-BR:Tchéquia ru:Чешская Республика zh-CN:捷克共和国]} RepresentedCountry:{GeoNameID:0 IsInEuropeanUnion:false IsoCode: Names:map[] Type:} Traits:{IsAnonymousProxy:false IsSatelliteProvider:false}}
[engine] mmdblookup: ASN: &{AutonomousSystemNumber:6830 AutonomousSystemOrganization:Liberty Global B.V.}
[engine] jsonapi: method: GET
[engine] jsonapi: URL: https://bouncer.ooni.io/api/v1/collectors
[engine] dns: resolving bouncer.ooni.io
[engine] dns: got 1 entries
[engine] - 37.218.245.90
[engine] connect: using tcp, 37.218.245.90:443
[engine] connect: connected to tcp, 37.218.245.90:443
[engine] tls: starting handshake
[engine] tls: handshake OK
[engine] - negotiated protocol: 
[engine] - version: TLSv1.2
[engine] http: connection to 37.218.245.90:443 ready; sending request
[engine] > GET /api/v1/collectors HTTP/1.1
[engine] > User-Agent: ooniprobe-cli/3.0.0-rc.9
[engine] > Accept-Encoding: gzip
[engine] > Host: bouncer.ooni.io
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Content-Type: application/json
[engine] < Connection: keep-alive
[engine] < Keep-Alive: timeout=120
[engine] < Etag: W/"306a5809d83abe223b2be2295ca750478cb8ca48"
[engine] < Server: nginx
[engine] < Date: Wed, 12 Feb 2020 10:55:50 GMT
[engine] <
[engine] jsonapi: response body: [{"type": "onion", "address": "httpo://jehhrikjjqrlpufu.onion"}, {"type": "https", "address": "https://ams-ps2.ooni.nu:443"}, {"front": "dkyhjv0wpi2dk.cloudfront.net", "type": "cloudfront", "address": "https://dkyhjv0wpi2dk.cloudfront.net"}, {"type": "onion", "address": "httpo://hcn5nqahdkds6cjv.onion"}, {"type": "https", "address": "https://mia-ps2.ooni.nu:443"}, {"front": "dkyhjv0wpi2dk.cloudfront.net", "type": "cloudfront", "address": "https://dkyhjv0wpi2dk.cloudfront.net"}]
[engine] jsonapi: method: GET
[engine] jsonapi: URL: https://bouncer.ooni.io/api/v1/test-helpers
[engine] http: connection to 37.218.245.90:443 ready; sending request
[engine] > GET /api/v1/test-helpers HTTP/1.1
[engine] > Host: bouncer.ooni.io
[engine] > User-Agent: ooniprobe-cli/3.0.0-rc.9
[engine] > Accept-Encoding: gzip
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Content-Type: application/json
[engine] < Connection: keep-alive
[engine] < Keep-Alive: timeout=120
[engine] < Etag: W/"77a124db5e44b42788180107dd0c84b7da83e03e"
[engine] < Server: nginx
[engine] < Date: Wed, 12 Feb 2020 10:55:50 GMT
[engine] <
[engine] jsonapi: response body: {"traceroute": [{"type": "legacy", "address": "37.218.241.93"}, {"type": "legacy", "address": "37.218.241.93"}], "ssl": [{"type": "legacy", "address": "https://37.218.241.93"}, {"type": "legacy", "address": "https://37.218.241.93"}], "web-connectivity": [{"type": "legacy", "address": "httpo://o7mcp5y4ibyjkcgs.onion"}, {"type": "https", "address": "https://wcth.ooni.io"}, {"front": "d33d1gs9kpq1c5.cloudfront.net", "type": "cloudfront", "address": "https://d33d1gs9kpq1c5.cloudfront.net"}, {"type": "legacy", "address": "httpo://y3zq5fwelrzkkv3s.onion"}, {"type": "https", "address": "https://wcth.ooni.io"}, {"front": "d33d1gs9kpq1c5.cloudfront.net", "type": "cloudfront", "address": "https://d33d1gs9kpq1c5.cloudfront.net"}], "dns": [{"type": "legacy", "address": "37.218.241.93:57004"}, {"type": "legacy", "address": "37.218.241.93:57004"}], "http-return-json-headers": [{"type": "legacy", "address": "http://37.218.241.94:80"}, {"type": "legacy", "address": "http://37.218.241.94:80"}], "tcp-echo": [{"type": "legacy", "address": "37.218.241.93"}, {"type": "legacy", "address": "37.218.241.93"}]}
   • Running test group Instant Messaging
   • Creating result {0 im 2020-02-12 10:55:50.996373655 +0000 UTC 17 0 false false 0 0 /root/.ooniprobe/msmts/im-2020-02-12T105550.996373655Z}
   ⨯ panic in app.Run          error=runtime error: invalid memory address or nil pointer dereference
   • sending exception backtrace

```

Reply:
@bassosimone Ok, so I replaced default sentry url with my own url and get the same error as you.

Reply:
I got this error when testing in malta-be (but not armvirt-64, although I didn't let it finish the whole test). I don't have a stack trace to compare though.

Perhaps you can try running upstream's pre-built (static) binary to see if you get the same error? If it doesn't happen with their binary then it's probably something in our build.

Reply:
I just tested version rc9 to rc3. The same error is in 9,8,7,6,5,4 but version 3 works as expected.

Reply:
Just want to confirm: Did you also change the version of probe-engine used when testing the different versions?

Reply:
Yes (more or less)-  I removed the part with probe-engine, since after your changes to golang its no longer necessary.  I will push updated package.